nemo/collections/duplex_s2s/models/duplex_s2s_model.py

```diff
@@ -654,6 +617,424 @@ def configure_model(self) -> None:
+                    tokens=gen_audio_codes.transpose(1, 2), tokens_len=lengths
+                )
+
+            asr_hyps = self.asr.transcribe(
+                [
+                    audio[:alen]
+                    for audio, alen in zip(
+                        torchaudio.functional.resample(predicted_audio, 22050, 16000),
+                        predicted_audio_lens,
+                    )
+                ],
+                batch_size=predicted_audio.shape[0],
+                verbose=False,
+            )
+
+            txt_hyps = [
+                self.tokenizer.ids_to_text(hyp_ids[:hyp_len]) for hyp_ids, hyp_len in zip(gen_text.cpu(), lengths)
+            ]
+
+
+            for ref, txt_hyp, asr_hyp in zip(dataset_batch["target_texts"], txt_hyps, asr_hyps):
+                asr_hyp = asr_hyp.text
+                self._refs[name].append([self.normalizer(ref)])
+                self._txt_preds[name].append(self.normalizer(txt_hyp))
+                self._asr_preds[name].append(self.normalizer(asr_hyp))
+                txtb = sacrebleu.sentence_bleu(txt_hyp, [ref]).score
+                asrb = sacrebleu.sentence_bleu(asr_hyp, [ref]).score
+                logging.info(f"[REF]\t{ref}\n[HYP]\t{txt_hyp} [{txtb:.2f}]\n[ASR]\t{asr_hyp} [{asrb:.2f}]")
+
+
+    @torch.no_grad()
+    def autoregressive_decoding(
+            self,
+            input_signal: Tensor,
+            input_signal_lens: Tensor
+    ) -> tuple[Tensor, Tensor, Tensor]:
+        """
+        Autoregressive prediction, compatible with both FSDP and non-FSDP modes.
+        If using FSDP (self._use_fsdp=True):
+        (1) Synchronize max sequence length across ranks
+        (2) Pad to that max length and run AR decoding
+        (3) Trim outputs back to local length
+        Without these steps, deadlock would happen during AR decoding (FSDP only)
+
+        """
+
+        input_embeds, lengths = self.perception(
+            input_signal=input_signal,
+            input_signal_length=input_signal_lens,
+        )
+        B, T_local, H = input_embeds.shape
+
+        # Determine decoding length and pad if FSDP
+        fsdp = getattr(self, '_use_fsdp', False)
+        if fsdp:
+            T_tensor = torch.tensor([T_local], device=input_embeds.device)
+            dist.all_reduce(T_tensor, op=dist.ReduceOp.MAX)
+            T = int(T_tensor.item())
+            if T > T_local:
+                last_frame = input_embeds[:, T_local - 1:T_local, :]  # (B,1,H)
+                pad = last_frame.repeat(1, T - T_local, 1)  # (B, T-T_local, H)
+                input_embeds = torch.cat([input_embeds, pad], dim=1)
+        else:
+            T = T_local
+
+        # Apply channel weight
+        input_embeds *= self.cfg.get("duplex_user_channel_weight", 1.0)
+
+        # This cache is for self.llm
+        cache = DynamicCache()
+        # Call reset_input_and_kv_cache to enable cache for TransformerARSpeechDecoder
```

Comment:
it would be good if TransformerARSpeechDecoder had explicit Cache argument that can be managed by the user.